### ISSUE TITLE
Problem when executing validation callback

### ISSUE THREAD

**Issue by jmartinpizarro** [ultralytics/ultralytics#22936]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am trying to implement a logger for the MAE of x,y coordinates for each epoch in the validation (and test, but that's another thing...) stage. With my current code, I obtain the error:

'DetectionValidator' object has no attribute 'predictions'

Which makes sense, because in the documentation there is no such attribute. However, I don't know if I am able to access to those values. Is it possible. Am I implemething this wrong? 

Help is appreciated
```python
def on_val_batch_end(validator):
    """
    validator: DetectionValidator
    """

    # preds: list of tensors (N, 6) → [x,y,w,h,conf,cls]
    preds = validator.predictions  # tensor
    if preds is None or len(preds) == 0:
        return

    # Ground truth
    gts = validator.batch["bboxes"]  # (N,4) -> x,y,w,h

    # Centros (x,y)
    preds_xy = preds[:, :2]
    gts_xy = gts[:, :2]

    # Distancia euclídea
    dist = torch.norm(preds_xy - gts_xy, dim=1)

    acc_xy = (dist < 0.02).float().mean()
    mae_xy = dist.mean()

    # Inicializar acumuladores si no existen
    if not hasattr(validator, "acc_xy"):
        validator.acc_xy = []
        validator.mae_xy = []

    validator.acc_xy.append(acc_xy.item())
    validator.mae_xy.append(mae_xy.item())
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jmartinpizarro, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This looks like a ⚠️ potential bug / API mismatch around validator callback expectations (e.g., what attributes are available on `DetectionValidator` during `on_val_batch_end`). To help us reproduce and confirm the expected callback interface, please share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE). 🙏

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---  

🤖 This is an automated response to help us triage quickly. An Ultralytics engineer will also assist soon. In the meantime, please paste your MRE plus the exact command you run (CLI or Python), your `ultralytics` version, and the full stack trace showing where `validator.predictions` is accessed so we can pinpoint the correct callback data flow.

**Comment by jmartinpizarro** [comment]
- **Ultralytics version**:  8.3.225

**Code were the problem can be found**:

```python
"""
This script contains the code for visualizing and computing accuracy metrics for the YOLO predictions (and its modifications)
"""

import os
import argparse

import pandas as pd
from ultralytics import YOLO

TESTING_EXPORT_DIR = 'model_testing_results'


def get_args():
    parser = argparse.ArgumentParser(
        description='YOLO and YOLO-mods Results Renderer'
    )

    parser.add_argument(
        '--data',
        default='data/yolo_v3/data.yaml',
        type=str,
        help='Route where the .yaml definition of the dataset for YOLO can be found.',
    )

    parser.add_argument(
        '--results-folder',
        default='runs',
        type=str,
        help="Route where the output model and its results are generated. By default is 'runs/'",
    )

    return parser.parse_args()


def main():
    args = get_args()

    best_precision = 0.0
    best_precission_model = None
    best_map50 = 0.0
    best_map50_model = None

    models = os.listdir(args.results_folder)

    models_output = {}

    print('[process_yolo_results] :: Starting the script\n')

    for model in models:
        print(f'\n\n\t[process_yolo_results] :: Testing {model}\n\n')

        route = os.path.join(args.results_folder, model)
        models_output[model] = {}

        # if the model could not be trained, skip it
        if model == 'detect' or not os.listdir(os.path.join(route, 'weights')):
            continue

        # now it is necessary to obtain the metrics with the test set
        YOLO_MODEL = YOLO(f'{route}/weights/best.pt')
        prediction_metrics = YOLO_MODEL.val(
            data=args.data,
            split='test',
            imgsz=640,
            half=True,
            device='cuda',
            save=True,
            name=route,
            save_conf=True,
            save_txt=False,
            save_json=False,
            verbose=False,
        )
        
        if prediction_metrics.box.p[0] >= best_precision:
            best_precision = prediction_metrics.box.p[0]  # only one class
            best_precission_model = model

        if prediction_metrics.box.map50 >= best_map50:
            best_map50 = prediction_metrics.box.map50
            best_map50_model = model

        print(
            '[process_yolo_results] :: The script has ended\n',
            f'\t The model with the best precission was: {best_precission_model} with {best_precision}\n',
            f'\t The model with the best map50 was: {best_map50_model} with {best_map50}\n',
        )

if __name__ == '__main__':
    main()

```

**Comment by Y-T-G** [comment]
Validator doesn't store ground truth and predictions as attributes

https://github.com/ultralytics/ultralytics/blob/532459b59c30a446d82f4e13942d054782497ef4/ultralytics/engine/validator.py#L208

**Comment by Y-T-G** [comment]
You will have to modify source code to add your metric

**Comment by jmartinpizarro** [comment]
Could you clarify that? Shall I just modifiy the source code (from the dependency I guess) from the validator?

**Comment by jmartinpizarro** [comment]
I added a custom metric rmse_xy inside DetectionValidator (models/yolo/detect/val.py). It is computed per-image during validation, accumulated across the epoch, and finalized at the end. Training/validation runs fine, but rmse_xy never appears in results.csv.

What I changed:

Initialize accumulators in init_metrics():
```python
self.rmse_sum = torch.tensor(0.0, device=self.device)
self.rmse_n = 0
self.rmse_xy = torch.tensor(0.0, device=self.device)
```

In update_metrics(), match predictions to GT with IoU and accumulate squared error of bbox centers:
```python
iou = box_iou(pbatch["bboxes"], predn["bboxes"])   # [N_gt, N_pred]
best_iou, best_idx = iou.max(dim=1)
valid = best_iou > 0.5

gt_idx = torch.arange(iou.shape[0], device=iou.device)[valid]
pred_idx = best_idx[valid]

# remove duplicate preds (keep highest IoU)
unique = {}
for g, p, i in zip(gt_idx.tolist(), pred_idx.tolist(), best_iou[valid].tolist()):
    if p not in unique or i > unique[p][1]:
        unique[p] = (g, i)
if not unique:
    continue

pred_idx = torch.tensor(list(unique.keys()), device=iou.device)
gt_idx = torch.tensor([v[0] for v in unique.values()], device=iou.device)

pb = predn["bboxes"][pred_idx]
gb = pbatch["bboxes"][gt_idx]

pcx = (pb[:, 0] + pb[:, 2]) / 2
pcy = (pb[:, 1] + pb[:, 3]) / 2
gcx = (gb[:, 0] + gb[:, 2]) / 2
gcy = (gb[:, 1] + gb[:, 3]) / 2

dx2 = (pcx - gcx) ** 2
dy2 = (pcy - gcy) ** 2

self.rmse_sum += dx2.sum() + dy2.sum()
self.rmse_n += len(pred_idx) * 2
```

Compute final RMSE in finalize_metrics():
```python
self.rmse_xy = torch.sqrt(self.rmse_sum / max(self.rmse_n, 1))
```

Tried to add it to the returned stats by injecting it in get_stats():
```python
self.metrics.process(save_dir=self.save_dir, plot=self.args.plots, on_plot=self.on_plot)
self.metrics.results_dict["rmse_xy"] = float(self.rmse_xy)
self.metrics.clear_stats()
return self.metrics.results_dict
```

Result: rmse_xy still does not appear in results.csv.

I think the metric is computed correctly, but the training logger/CSV writer is not using metrics.results_dict (or is filtering keys). What is the recommended way to add a custom scalar metric so it is included in results.csv during training?

**Comment by glenn-jocher** [comment]
You’re very close, but `rmse_xy` is being computed in `finalize_metrics()`, while the value that gets logged/written comes from the dict returned by `get_stats()` (called earlier in the validation loop), so nothing new ever makes it into the returned `stats` dict (see the call order in [`BaseValidator.__call__`](https://docs.ultralytics.com/reference/engine/validator/#__call__)). Also, `self.metrics.results_dict` is best treated as read-only (it’s generated), so mutating it in-place can be ignored.

Compute `rmse_xy` inside `get_stats()` and return it as part of the dict (matching the existing `metrics/...` naming used by detection), e.g. by overriding [`DetectionValidator.get_stats`](https://docs.ultralytics.com/reference/models/yolo/detect/val/#ultralytics.models.yolo.detect.val.DetectionValidator.get_stats):

```python
def get_stats(self):
    self.metrics.process(save_dir=self.save_dir, plot=self.args.plots, on_plot=self.on_plot)
    stats = dict(self.metrics.results_dict)  # copy
    rmse_xy = torch.sqrt(self.rmse_sum / max(self.rmse_n, 1)).item()
    stats["metrics/rmse_xy(B)"] = rmse_xy
    self.metrics.clear_stats()
    return stats
```

If you’re training with DDP, you’ll also need to `all_reduce`/sum `rmse_sum` and `rmse_n` across ranks in `gather_stats()`, otherwise rank-0 logging won’t reflect the full epoch.

**Comment by jmartinpizarro** [comment]
Thanks for the help, the problem it is now solved. I have a couple of questions just to confirm minor details:

1. For plotting the F1-Score, I want to use the RMSE instead of accuracy (due to the problem I am doing, accuracy is not really that relevant). Is it possible to change it? 
2. Wouldn't be relevant to include metrics such as MSE, RMSE... Those types of errors in the general outputs of YOLO?

**Comment by glenn-jocher** [comment]
Glad it’s working. The built-in F1/PR plots are derived from precision/recall vs confidence (not “accuracy”) and are generated inside the detection metrics pipeline, so there isn’t a supported switch to replace those curves with RMSE; the practical approach is to log your `metrics/rmse_xy(B)` scalar and plot it separately from `results.csv`, or extend the plotting code in `ultralytics/utils/metrics.py` to add an extra RMSE plot alongside the defaults (the metric aggregation pattern is shown in [`DetMetrics.update_stats()`](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.DetMetrics.update_stats)).  

We don’t ship MSE/RMSE by default because they’re highly application-specific (matching policy, normalization/units, what error is meaningful), but adding them as a custom scalar like you did is the recommended path; if you think an optional “localization error” metric would be broadly useful, please open a feature request in the repo (or a small PR) so we can discuss the expected definition and naming.

**Comment by jmartinpizarro** [comment]
Thanks for the help!

**Comment by glenn-jocher** [comment]
You’re welcome—if you decide you want this RMSE-style metric supported more broadly, please open a feature request with the exact definition you’re using in [Ultralytics Discussions](https://github.com/orgs/ultralytics/discussions) (or a new issue) so it’s easier for the team to review.
